docs/rfcs/20220829-distributed-dynamic-routing.md

```diff
@@ -0,0 +1,224 @@
+ ┌──┤Node0 ├──┐  ┌──┤Node1 ├──┐   ┌──┤Node2 ├──┐ │       │ ┌──┤Node0 ├──┐  ┌──┤Node1 ├──┐   ┌──┤Node2 ├──┐ 
+ │  └──────┘  │  │  └──────┘  │   │  └──────┘  │ │ Table │ │  └──────┘  │  │  └──────┘  │   │  └──────┘  │ 
+ │┌────────┐  │  │            │   │┌────────┐  │ │ Load  │ │┌────────┐  │  │            │   │┌────────┐  │ 
+ ││Shard0 L├┐ │  │┌─────────┐ │   ││Shard1 L├┐ │ │Balance│ ││Shard0 L├┐ │  │┌─────────┐ │   ││Shard1 L├┐ │ 
+ │├────────┘│ │  ││Shard0 F │ │   │├────────┘│ │ │       │ │├────────┘│ │  ││Shard0 F │ │   │├────────┘│ │ 
+ ││ table0  │ │  │└─────────┘ │   ││ table2  │ │ └───────┘ ││ table0  │ │  │└─────────┘ │   ││ table1  │ │ 
+ ││ table1  │ │  │┌─────────┐ │   ││ table3  │ │──────────▶││         │ │  │┌─────────┐ │   ││ table2  │ │ 
+ ││         │ │  ││Shard1 F │ │   ││         │ │           ││         │ │  ││Shard1 F │ │   ││ table3  │ │ 
+ │└─────────┘ │  │└─────────┘ │   │└─────────┘ │           │└─────────┘ │  │└─────────┘ │   │└─────────┘ │ 
+ │┌─────────┐ │  │            │   │┌─────────┐ │           │┌─────────┐ │  │            │   │┌─────────┐ │ 
+ ││Shard2 L │ │  │            │   ││Shard2 F │ │           ││Shard2 L │ │  │            │   ││Shard2 F │ │ 
+ │└─────────┘ │  │            │   │└─────────┘ │           │└─────────┘ │  │            │   │└─────────┘ │ 
+┌┴────────────┴┐ └────────────┘  ┌┴────────────┴┐         ┌┴────────────┴┐ └────────────┘  ┌┴────────────┴┐
+│  CPUUse=90%  │                 │  CPUUse=30%  │         │  CPUUse=50%  │                 │  CPUUse=50%  │
+└──────────────┘                 └──────────────┘         └──────────────┘                 └──────────────┘
+```
+
+### Shard Change Process
+#### Switch Leader
+
+1. First, set the original leader shard and the follwer to be switched to PENDING_ FOLLOWER and PENDING_ LEADER.
+1. Wait for the leader shard to complete the preparatory action before the leader switch, and notify the CeresMeta after the completion. In this process, the new leader cannot process the write request.
+1. After receiving this request, meta will update the status of the two shards, and the leader switch is completed.
+
+
+#### Shard Migrate
+
+- Leader Shard：
+    1. First, CeresMeta adjusts the routing relationship and create a new follower shard on the node.
+    1. Follower shard notifies CeresMeta after initialization.
+    1. CeresMeta started the process of leader switch, switched the leader to the newly created follower shard.
+    1. Close the original leader shard after the leader switch is completed.
+- Follower Shard：
+    1. Adjust shard node mapping to directly assign this shard to another node.
+
+#### Shard Split
+
+1. CeresMeta creates a new shard and distributes it to the node with the lowest load.
+1. After the new shard is created, the table routing relationship in the shard is processed, and a part of the table is split to the new shard according to certain rules.
+
+Pending Problem：
+
+1. How to decide which tables need to be split into new shards?
```

Comment:
Another thing worth mentioning:  `Shard` is consisted of a group of `Table`s. Switching `Shard`'s role is actually switching lots of `Table`s in Instance, it's not an atomic operation. What if some table failed to switch its role? Should CeresMeta move them out of that shard or do something else?

Reply:
This is an important problem, and it requires some discussion for a solution to it.

Reply:
> Another thing worth mentioning: `Shard` is consisted of a group of `Table`s. Switching `Shard`'s role is actually switching lots of `Table`s in Instance, it's not an atomic operation. What if some table failed to switch its role? Should CeresMeta move them out of that shard or do something else?

This is also what I have not considered clearly at present, I will update it after discussion clearly